### PR TITLE
Bugfix/caching public key

### PR DESCRIPTION
**Pull request by algoidan**

### CODE REVIEW

File: src/algo_addr.c [hunk 1]
```diff
@@ -12,7 +12,7 @@ checksummed_addr(const uint8_t *publicKey, char *out)
   // The SDK does not provide a ready-made SHA512/256, so we set up a SHA512
   // hash context, and then overwrite the IV with the SHA512/256-specific IV.
   cx_sha512_t h;
-  memset(&h, 0, sizeof(h));
+  explicit_bzero(&h, sizeof(h));
```
Comment (algorotem) [inline src/algo_addr.c:15]: I like this change.

**Comment by algorotem** [comment]
Will continue to review later.

File: src/algo_tx_dec.c [hunk 2]
```diff
@@ -61,8 +61,7 @@ decode_string(uint8_t **bufp, uint8_t *buf_end, char *strbuf, size_t strbuflen)
     snprintf(decode_err, sizeof(decode_err), "expected string, found %d", b);
     THROW(INVALID_PARAMETER);
   }
-
-  if (str_len > strbuflen) {
+  if (str_len >= strbuflen) {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply src/algo_tx_dec.c:64]: if str_len is equal to strbuflen, the function will write '\0' outside the bounds of the strbuf.

File: src/algo_keys.c [hunk 3]
```diff
@@ -5,31 +5,44 @@
...
-  os_perso_derive_node_bip32(CX_CURVE_Ed25519, bip32Path, sizeof(bip32Path) / sizeof(bip32Path[0]), privateKeyData, NULL);
-  cx_ecfp_init_private_key(CX_CURVE_Ed25519, privateKeyData, 32, privateKey);
-
+  BEGIN_TRY
+  {
+    TRY
+    {
+      os_perso_derive_node_bip32(CX_CURVE_Ed25519, bip32Path, sizeof(bip32Path) / sizeof(bip32Path[0]), private_key_data, NULL);
+      cx_ecfp_init_private_key(CX_CURVE_Ed25519, private_key_data, 32, &local_private_key);
+      os_memcpy(private_key, &local_private_key, sizeof(local_private_key));
```
[the remark this answers is not in the document]
Comment (algoidan) [reply src/algo_keys.c:33]: 1 - according to the documentation, the function cx_ecfp_init_private_key inits the struct cx_ecfp_private_key_t with the raw private key derived from os_perso_derive_node_bip32. this struct is used to call other APIs (such as sign).
2 - This cleanup technique is recommended only when handling sensitive data. zeroing other fields if not necessary, might have an impact on the app performances.

File: src/algo_addr.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (algoidan) [reply src/algo_addr.c:14]: :)

File: src/algo_keys.c [hunk 4]
```diff
@@ -43,25 +65,72 @@ algorand_public_key(const cx_ecfp_private_key_t *privateKey, uint8_t *buf)
...
+  {
+    TRY
+    {
+      algorand_key_derive(account_id, &private_key);
+      get_public_key_from_private_key(&private_key, out_pub_key, out_pub_key_size);
+    }
+    FINALLY
+    {
+      explicit_bzero(&private_key, sizeof(private_key));
+    }
```
Comment (algorotem) [inline src/algo_keys.c:97]: I'm probably missing something but this try catch is very confusing for me. 
Assuming `algorand_key_derive` failed, you'll have garbage in `out_pub_key` and you won't know about it. Isn't it suppose to throw something if it fails? or do we just suppress the error?

File: src/command_handler.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
+#include "command_handler.h"
+#include "apdu_protocol_defines.h"
+#include "algo_ui.h"
+#include "algo_addr.h"
+#include "algo_keys.h"
+
+
+
+
+void parse_input_get_public_key(const uint8_t* buffer, const uint32_t buffer_len, uint32_t* output_account_id )
```
Comment (algorotem) [inline src/command_handler.c:10]: remove the excessive space at the end just for my ocd

File: src/command_handler.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+
+/*
+* This function takes a binary 32 bytes public key, converts it to Algorand public address,
+* and send it to the UI.
+* this function fails if the public_key buffer is small than 32 bytes.
+*/
+void send_address_to_ui(const uint8_t* public_key, const uint32_t public_key_size)
+{
+  if (public_key_size < ALGORAND_PUBLIC_KEY_SIZE)
```
Comment (algorotem) [inline src/command_handler.c:48]: Shouldn't it be strictly equal (==)? larger is also no good.

File: src/command_handler.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    THROW(0x6a87);
+  } 
+  *output_account_id = U4BE(buffer, OFFSET_CDATA);
+}
+
+
+/*
+* This function takes a binary 32 bytes public key, converts it to Algorand public address,
+* and send it to the UI.
+* this function fails if the public_key buffer is small than 32 bytes.
```
Comment (algorotem) [inline src/command_handler.c:44]: nit - `Algorand address` and not `Algorand public address`
typo - small -> smaller

File: src/main.c [hunk 8]
```diff
@@ -6,54 +6,37 @@
...
  */
+
 #if defined(TARGET_NANOX)
-static uint8_t msgpack_buf[2048];
+#define TNX_BUFFER_SIZE 2048
+static uint8_t msgpack_buf[TNX_BUFFER_SIZE];
 #else
-static uint8_t msgpack_buf[900];
+#define TNX_BUFFER_SIZE 900
+static uint8_t msgpack_buf[TNX_BUFFER_SIZE];
```
Comment (algorotem) [inline src/main.c:30]: lines 27 and 30 are duplicative, consider taking them out of the if to avoid that.

File: src/main.c [hunk 9]
```diff
@@ -219,90 +200,40 @@ algorand_main(void)
...
           uint8_t user_approval_required = G_io_apdu_buffer[OFFSET_P1] == P1_WITH_REQUEST_USER_APPROVAL;
-
-          if (rx > OFFSET_LC) {
-              uint8_t lc = G_io_apdu_buffer[OFFSET_LC];
-              if (lc == sizeof(uint32_t)) {
-                accountId = U4BE(G_io_apdu_buffer, OFFSET_CDATA);
-              } else if (lc != 0) {
-                return THROW(0x6a85);
-              }
-          }
```
Comment (algorotem) [inline src/main.c:291]: How do users get non-zero index addresses?

File: tests/test_utils/get_public_key.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+
+from ledgerblue.comm import getDongle
+import struct
+
+
+if __name__ == '__main__':
+    dongle = getDongle(True)
+    apdu = struct.pack('>BBBBBI', 0x80, 0x3, 0x80, 0x0, 0x0, 0x0)
+    dongle.exchange(apdu)
```
Comment (rotemh) [inline tests/test_utils/get_public_key.py:9]: what's the purpose of this code?

**Pull request merged**